### PR TITLE
support builtin for Dir class singleton method

### PR DESCRIPTION
**Pull request by S-H-GAMELINKS**
support builtin for Dir class these singleton method's.

```
Dir#foreach
Dir#entries
Dir#each_child
Dir#children
```

### CODE REVIEW

File: dir.rb [hunk 1]
```diff
@@ -24,6 +24,97 @@ def self.open(name, encoding: nil, &block)
...
+  #  <em>produces:</em>
+  #
+  #     Got .
+  #     Got ..
+  #     Got config.h
+  #     Got main.rb
+  #
+  #
+  def self.foreach(path, encoding: Encoding.find("filesystem"))
+    return Enumerator.new(Dir.new(path, encoding: encoding)) if !block_given?
```
Comment (nobu) [inline dir.rb:50]: `RETURN_ENUMERATOR` defers to open `path`.

### master
```shell
$ rm -fr xxx; ruby -e 'd = Dir.foreach("xxx"); Dir.mkdir("xxx"); p d.to_a'
[".", ".."]
```

### patched
```shell
$ rm -fr xxx; ./ruby -e 'd = Dir.foreach("xxx"); Dir.mkdir("xxx"); p d.to_a'
Traceback (most recent call last):
	3: from -e:1:in `<main>'
	2: from <internal:dir>:50:in `foreach'
	1: from <internal:dir>:50:in `new'
<internal:dir>:126:in `initialize': No such file or directory @ dir_initialize - xxx (Errno::ENOENT)
```

Comment (nobu) [reply dir.rb:50]: Note that `Enumerator.new` yields a warning.

```shell
$ ruby -e 'Enumerator.new(Dir.new("."))'
-e:1: warning: Enumerator.new without a block is deprecated; use Object#to_enum instead
```

File: dir.c [hunk 2]
```diff
@@ -2955,42 +2955,33 @@ dir_s_glob(rb_execution_context_t *ec, VALUE obj, VALUE str, VALUE rflags, VALUE
...
     return dir;
 }
 
+static void
+dir_open_argv(VALUE *argv, VALUE path, VALUE enc)
+{
+    VALUE kw_hash = rb_hash_new();
+    rb_gc_register_mark_object(kw_hash);
+    rb_hash_aset(kw_hash, ID2SYM(rb_intern("encoding")), enc);
+    rb_obj_freeze(kw_hash);
```
Comment (nobu) [inline dir.c:2970]: The purpose to move C-methods to builtin methods is to eliminate overheads of hash creation for keywords argument.
Making a hash here loses it.

Comment (S-H-GAMELINKS) [reply dir.c]: Remove and Fix it.

File: dir.rb (hunk 1 above)
Comment (S-H-GAMELINKS) [reply dir.rb]: Change code to using `RETURN_ENUMERATOR`.

Comment (nobu) [reply dir.rb]: Passing `path` only makes `encoding:` option not working.
The enumerator should be returned in dir.rb, I guess.

Comment (S-H-GAMELINKS) [reply dir.rb]: OK.
Fix this patch, like this.

```c
    VALUE argv[2];
    argv[0] = path;

    if (NIL_P(enc)) {
    RETURN_ENUMERATOR(io, 1, argv);
    } else {
    VALUE kw_hash = rb_hash_new();

    rb_gc_register_mark_object(kw_hash);
    rb_hash_aset(kw_hash, ID2SYM(rb_intern("encoding")), enc);
    rb_obj_freeze(kw_hash);
    argv[1] = kw_hash;

    RETURN_ENUMERATOR_KW(io, 2, argv, RB_PASS_KEYWORDS);
    }
```

Comment (S-H-GAMELINKS) [reply dir.rb]: create `kw_hash`,  workaround for  `ArgumentError` 

```bash
$ rm -fr xxx; ruby -e 'd = Dir.foreach("xxx"); Dir.mkdir("xxx"); p d.to_a'
Traceback (most recent call last):
        2: from -e:1:in `<main>'
        1: from <internal:dir>:51:in `foreach'
<internal:dir>:15:in `open': No such file or directory @ dir_initialize - xxx (Errno::ENOENT)
```

**Pull request closed without merging**